torch/csrc/distributed/rpc/request_callback.cpp

```diff
@@ -28,16 +28,24 @@ struct ClearAutogradContextGuard {
 
 } // anonymous namespace
 
-Message RequestCallback::operator()(Message& request) const {
+std::shared_ptr<FutureMessage> RequestCallback::operator()(
+    Message& request) const {
   // For a recv thread, current context id should be invalid outside
   // processMessage().
   ClearAutogradContextGuard guard;
   try {
-    return processMessage(request);
+    // do we need to run clear() in the future invocation?
+    auto ret = processMessage(request);
+    if (!ret->completed()) {
+      // If not yet complete, clear the context when we do.
+      ret->addCallback(
```

Comment:
We aren't using thread-local for context, are we?
 - I did just tweak this to avoid clearing context inline in this !completed() case.
 - I don't necessarily expect that we will necessarily implement Python UDFs server-side in a pure async manner - though maybe this is reasonable to do.
 - We now won't try to clear until after the future message comes back... but not sure if this is sufficient